src/libcore/ptr.rs

```diff
@@ -2556,47 +2549,79 @@ impl<T: ?Sized> NonNull<T> {
     /// The resulting lifetime is bound to self so this behaves "as if"
     /// it were actually an instance of T that is getting borrowed. If a longer
     /// (unbound) lifetime is needed, use `&mut *my_ptr.as_ptr()`.
-    #[stable(feature = "nonnull", since = "1.24.0")]
+    #[stable(feature = "nonnull", since = "1.25.0")]
     pub unsafe fn as_mut(&mut self) -> &mut T {
         &mut *self.as_ptr()
     }
+
+    /// Cast to a pointer of another type
+    #[unstable(feature = "nonnull", issue = "27730")]
+    pub fn cast<U>(self) -> NonNull<U> {
```

Comment:
Shouldn't this be `U: ?Sized`?

Reply:
It should. I’ll also add some tests.

Reply:
… or not:

```
error[E0606]: casting `*mut T` as `*mut U` is invalid
     = note: vtable kinds may not match
```

Reply:
Then maybe this method should also require `T: Sized`?